feat(RotationSystem): Deep Learning Support

1zun4 opened the pull request:
## DeepLearning Rotation System

The successor to https://github.com/CCBlueX/LiquidBounce/pull/5642 - which was far from finished, and unfortunately people tried to use it before I had it ready, so they asked for support and gave feedback on an unfinished pull request.

I am now looking for people who can play and record their combat data - in order to have enough for a ready to use model. I have prepared three models myself within 20 minutes total.

## Features
- [x] Deep Learning Framework
- [x] Download Progressbar
- [x] Rotation Model
- [x] Model Managment Command
- [x] Training Recorder
- [x] Combat Recorder
- [x] Base Model

### For the future
- [ ] Curve Recorder
- [ ] Model Configuration
- [ ] Parameter Configuration
- [ ] Model Metadata
- [ ] Model Load On Select

## Data Collection

### Minarai Combat (PTB)
![image](https://github.com/user-attachments/assets/354be75d-d3a9-49e7-a842-8bff252fc8b8)

No Recording. Imagine playing Practice, over... and over... and over...


### Minarai Trainer (PTP)
![image](https://github.com/user-attachments/assets/3bb2cc8c-bf3c-484a-b1fe-303c9dda8051)

https://github.com/user-attachments/assets/b0745224-c8e1-4b73-91d9-d6b9471722f2


## Model Managment
- `.models create <name>`
- `.models improve <name>` (!WARNING! This will not filter previously used data, remove it before improving!)
- `.models delete <name>`

![image](https://github.com/user-attachments/assets/4395960d-ae5e-4962-bbbe-f9743b6459bb)


## Roation Configuration

![image](https://github.com/user-attachments/assets/c53813c1-63e6-4b65-80d6-3733262dbf3d)

In case our model is not able to keep up with high movement, we can use `Correction` which will act as Aim Assist for the Model. This is very useful as e.g. when only training with PTB-Mode we end up with a lot of low-delta data causing our model to move too slowly. This is NOT an issue with PTP-Mode though. We can also amplify the model output, which has a similar effect and makes it much faster, but it will often aim in the wrong places or start to shake as it tries to correct the wrong output on the next tick.

## Demo (PTP+PTB Model)

### Model and Data
[Minarai-Izuna-20min-Training.zip](https://github.com/user-attachments/files/18897442/Minarai-Izuna-20min-Training.zip)

### Video

https://github.com/user-attachments/assets/f89517ff-ad4f-453b-ad6b-5129aa6ba4d1

### In-Depth
As we can see, it sort of combines the concept of **Acceleration** as well as **LazyFlick** (known in Legacy). @mems01 has already done a good job of replicating this behaviour in traditional methods, and I still think it's probably the better solution compared to Deep Learning - but it has to rely on the developer to keep upgrading, updating and maintaining the curves, because as @xkeksi has already pointed out, they always have recognisable patterns. Deep learning is a pattern in itself, but it's a very large and dynamic pattern, and with enough data we can make it behave differently - and with just a little bit more data we can build a whole new pattern without writing a single bit of code.

### Explanation

The problem with this model is that we're relying on the LiquidBounce Aim Point Tracker - which will conflict with our PTB model, which wants to introduce its own point tracking. That's why it started to drift a bit - but overall this clip shows how it works, and from the current feedback I'm getting, it seems I've finally got it right.

### Feedback
![image](https://github.com/user-attachments/assets/999b3231-b339-497b-8b6a-8992f79ee4f1)
![image](https://github.com/user-attachments/assets/7dbd790f-f062-4024-964f-4b01c4ee1be1)
![image](https://github.com/user-attachments/assets/1bc7ff16-b614-47e7-9102-4b083b68c9b2)

Thank you very much. It's always nice to hear that something you've been working on for the last 5 days is working. Of course we still have to see how it works with anti-cheats, but I will leave that to the config creators and just provide an easy to use model system :)

## To reviewers
Do NOT submit any code feedback. Thank you.

1zun4 committed:
feat(deeplearn): implement framework with model training

1zun4 committed:
fix: missing lib

1zun4 committed:
feat(deeplearn): entity distance delta and age

1zun4 committed:
fix(deeplearn): training data age inconsistency
This is due to the fact that in a simulated environment there will be a different ratio of age to death than in a normal combat scenario.

1zun4 committed:
refactor(deeplearn): obsolete data generator

1zun4 committed:
fix(HUD/Hotbar): overlay message formatting

1zun4 committed:
feat(deeplearn): enhanced training and dataset

1zun4 committed:
fix(deeplearn/correction): none overwriting current

1zun4 committed:
fix(deeplearn/command): create, improve and delete command

1zun4 committed:
feat(deeplearn): measure prediction performance

twcult commented:
Hurttime may be superfluous in this case, but I could be wrong. Look at its priority in features

1zun4 commented:
> Hurttime may be superfluous in this case, but I could be wrong. Look at its priority in features

The data is being recorded - however, it's not passed into the model.

https://github.com/CCBlueX/LiquidBounce/blob/e445ca915fae368523b114410496ba82ed925b46/src/main/kotlin/net/ccbluex/liquidbounce/deeplearn/data/TrainingData.kt#L69-L90

sqlerrorthing reviewed (commented):
aww fuck izuna fuck me

sqlerrorthing commented:
@1zun4 do u know russian? (at least a little?)

vanvac commented:
dam ai aimbot is cazy (anticheats are cooked)

sqlerrorthing commented:
![image](https://github.com/user-attachments/assets/1db6b487-fdf3-478f-8de9-1350deae5885)
invalid char in progress bar (after percents and before progress bar bar)

1zun4 commented:
> ![image](https://github.com/user-attachments/assets/1db6b487-fdf3-478f-8de9-1350deae5885)
> invalid char in progress bar (after percents and before progress bar bar)

That is normal. It's a line break and only works using the LiquidBounce HUD.

sqlerrorthing commented:
лол хфывфхывхыфвх 


> > ![image](https://private-user-images.githubusercontent.com/148702857/415614876-1db6b487-fdf3-478f-8de9-1350deae5885.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAxMzY2NDksIm5iZiI6MTc0MDEzNjM0OSwicGF0aCI6Ii8xNDg3MDI4NTcvNDE1NjE0ODc2LTFkYjZiNDg3LWZkZjMtNDc4Zi04ZGU5LTEzNTBkZWFlNTg4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIyMVQxMTEyMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YjU0MTdhNDZmZDUzOGM1NDRiODliYjE0M2E5ZWIwNWU0ZjE1YjUzMjg4MTU2ZGI4Y2UzODEzN2ZjZjg4MTFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.txb-Ki3JAfA6ub2LiNQNidWLckOsc3spjaJSy2HdVUk)
> > invalid char in progress bar (after percents and before progress bar bar)
> 
> That is normal. It's a line break and only works using the LiquidBounce HUD.

check for it?

1zun4 commented:
> check for it?

Not relevant.

superblaubeere27 commented on build.gradle:
```diff
@@ -160,6 +160,13 @@ dependencies {
+    // Machine Learning
+    includeDependency 'ai.djl:api:0.31.1'
+    includeDependency "ai.djl.pytorch:pytorch-engine:0.31.1"
+//    runtimeOnly "ai.djl.mxnet:mxnet-engine:0.31.1"
+//    runtimeOnly "ai.djl.tensorflow:tensorflow-model-zoo:0.31.1"
```
![image](https://github.com/user-attachments/assets/4cd296ef-80da-4676-8ded-82ab0d28839f)

sqlerrorthing replied:
do not review a draft pr

superblaubeere27 replied:
> do not review a draft pr

that's... the joke

commandblock2 commented:
Holy *, I am speechless, never thought a tiny model like this could be perfect for this task
```
private fun createMlpBlock(outputs: Long) = SequentialBlock()
    .add(Linear.builder()
        .setUnits(128)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(64)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(32)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(outputs)
        .build())
```

You did make it happen and that's crazy. I think this is the exactly right step to go.
I am indeed thinking about a neural network in the game tho I would do it with the client sending gamedata to / receiving input from another process (python or whatever).

sqlerrorthing commented:
![image](https://github.com/user-attachments/assets/1aa0d28a-baf9-4109-9e44-6ca8023597d7)
FUCK
AN HOUR OF RECORDING THIS SHIT

sqlerrorthing commented:
@1zun4, record with 100k samples (again, but it was saved successfully this time.)
download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

MukjepScarlet commented:
@sqlerrorthing we probably need a more effecient way to save the record XD

MukjepScarlet commented:
Well I found a eaiser way to solve the save issue

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt:
```diff
@@ -43,31 +47,24 @@ object ModuleDebugRecorder : ClientModule("DebugRecorder", Category.MISC) {
                     file = folder.resolve("${baseName}_${idx++}.json")
                 }
 
                 file.writeText(publicGson.toJson(this.packets))
```
This can fix the save failure:
```kotlin
file.bufferedWriter().use {
    publicGson.toJson(this.packets, it)
}
```

sqlerrorthing replied:
do not review, and check [it](https://github.com/CCBlueX/LiquidBounce/pull/5672/commits/e81196c4dc71614dc8d99980b090827276a45306#diff-84d2bd893025a299a7a199637be011492db5c859cab6afc92ef8980c5133e573R73-R76)

MukjepScarlet replied:
Well that's another problem, we should make sure the `folder` exists before saving

sqlerrorthing replied:
we sure.
https://github.com/CCBlueX/LiquidBounce/blob/e81196c4dc71614dc8d99980b090827276a45306/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt#L71

1zun4 commented:
> @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

Thank you. That is with PTP Training (Minarai Trainer)?

sqlerrorthing commented:
> > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> 
> Thank you. That is with PTP Training (Minarai Trainer)?

yes

MukjepScarlet commented:
It will be good if we can share and get models on the incoming marketplace, I'm looking forward to it.

vanvac commented:
will this feature require a decent computer to run cause from what im seeing its a ai controlling the player and ai dont seem too easy to run on not good pc's

sqlerrorthing commented:
> will this feature require a decent computer to run cause from what im seeing its a ai controlling the player and ai dont seem too easy to run on not good pc's

i think not. maybe add a setting, like a toggle switch.
remember that ai doesn't need many neurons to work

liquidsquid1 commented:
this seems really interesting! i can do a 6 hour recording soon and upload the model here with minarai combat? ill just be playing pika xd

DrParanoya commented:
It looks way too legit

1zun4 committed:
feat(DebugRecorder): buffered writing

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

vanvac commented:
is this planned to release with lb 1.27.0 or its done when its done type of thing

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

1zun4 commented:
> @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

![image](https://github.com/user-attachments/assets/704ba4f7-1e52-4272-adb0-4a5bebc97330)
Next time please upload to GitHub instead. Also compress the JSON before uploading.

1zun4 committed:
fix(ConfigSystem): Load after client initialization

sqlerrorthing commented:
> > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> 
> ![image](https://private-user-images.githubusercontent.com/12410754/416358295-704ba4f7-1e52-4272-adb0-4a5bebc97330.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0NDE0NTAsIm5iZiI6MTc0MDQ0MTE1MCwicGF0aCI6Ii8xMjQxMDc1NC80MTYzNTgyOTUtNzA0YmE0ZjctMWU1Mi00MjcyLWFkYjAtNGE1YmViYzk3MzMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI0VDIzNTIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlOTIzMmEzMzBhYTY3Njk1YjQwNTUxOWQ3ZTZmMmRmZmYwNjAzNWYxYWI3M2U5MDY1NmUzNTg2OGI4YWMwMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sfk1V0Nf9Lmu82kmknaFqyfxa3ojUYDP79ABts00bZs) Next time please upload to GitHub instead. Also compress the JSON before uploading.

oh, but this file more than 10 MiB

1zun4 commented:
> > > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> > 
> > 
> > ![image](https://private-user-images.githubusercontent.com/12410754/416358295-704ba4f7-1e52-4272-adb0-4a5bebc97330.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0NDE0NTAsIm5iZiI6MTc0MDQ0MTE1MCwicGF0aCI6Ii8xMjQxMDc1NC80MTYzNTgyOTUtNzA0YmE0ZjctMWU1Mi00MjcyLWFkYjAtNGE1YmViYzk3MzMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI0VDIzNTIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlOTIzMmEzMzBhYTY3Njk1YjQwNTUxOWQ3ZTZmMmRmZmYwNjAzNWYxYWI3M2U5MDY1NmUzNTg2OGI4YWMwMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sfk1V0Nf9Lmu82kmknaFqyfxa3ojUYDP79ABts00bZs) Next time please upload to GitHub instead. Also compress the JSON before uploading.
> 
> oh, but this file more than 10 MiB

Not as ZIP.
[1yc28l82.zip](https://github.com/user-attachments/files/18951761/1yc28l82.zip) (4,1 MiB)

MukjepScarlet commented:
We can have a special file format for `List<TrainingData>`. But this will make the files unreadable.

sqlerrorthing commented:
> > > > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> > > 
> > > 
> > > ![image](https://private-user-images.githubusercontent.com/12410754/416358295-704ba4f7-1e52-4272-adb0-4a5bebc97330.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0NDE0NTAsIm5iZiI6MTc0MDQ0MTE1MCwicGF0aCI6Ii8xMjQxMDc1NC80MTYzNTgyOTUtNzA0YmE0ZjctMWU1Mi00MjcyLWFkYjAtNGE1YmViYzk3MzMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI0VDIzNTIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlOTIzMmEzMzBhYTY3Njk1YjQwNTUxOWQ3ZTZmMmRmZmYwNjAzNWYxYWI3M2U5MDY1NmUzNTg2OGI4YWMwMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sfk1V0Nf9Lmu82kmknaFqyfxa3ojUYDP79ABts00bZs) Next time please upload to GitHub instead. Also compress the JSON before uploading.
> > 
> > oh, but this file more than 10 MiB
> 
> Not as ZIP.
> [1yc28l82.zip](https://github.com/user-attachments/files/18951761/1yc28l82.zip) (4,1 MiB)
> 
> 

не подумал, сорянчик

1zun4 commented:
> We can have a special file format for `List<TrainingData>`. But this will make the files unreadable.

That is not an issue but compressed JSON does the job as well.

1zun4 committed:
feat(deeplearn): base minarai model

This is the same combat model I posted at the beginning with PTP and PTB training combined.

1zun4 committed:
feat(deeplearn/command): browse and reload command

This allows to quickly drag and drop another model into LiquidBounce.

1zun4 committed:
fix(deeplearn/command): reload translation

xkeksi commented:
: )

liquidsquid1 commented:
> this seems really interesting! i can do a 6 hour recording soon and upload the model here with minarai combat? ill just be playing pika xd

haven't been banned yet, wow this does really look realistic.

1zun4 commented:
Just to let you know. This pull request is basically done, but I need to work on a new startup task manager that notifies the user that a required file is being downloaded, at what speed, and how much % is completed, similar to what we already do with MCEF - but this time integrated into LiquidBounce instead. At the same time, we can also use this to speed up the startup time of LiquidBounce in the future, as I will be working on moving more of the client initialization to asynchronous. Fortunately, we can inject into https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/util/Utils.java using a mixin - this allows us to monitor the progress of a DJL download. Currently DJL just prints it to the console without providing access to a progress listener, so we need to integrate one ourselves.

I think I will do this in another pull request though, and when that is done, integrate it properly here.

sqlerrorthing commented:
separate the library's downloading menu in mcef, provide only api for download & the download progress.

create downloading tasks, download the mcef library & download the deeplearn library.

create the downloading menu in the client and asynchronous run tasks & render their progress

i see it

7GrandDadPGN commented:
3 hours of playing on gamster practice with no mitigation, the deep learning bypass go crazy 🗣️🗣️
(I did have to add my own sword hand check to aura though to play legit, but it was fun)
![image](https://github.com/user-attachments/assets/5d251fbf-6c9c-4f7b-a625-1a77f857e42e)

1zun4 committed:
Merge branch 'nextgen' into feat/deep-learning

# Conflicts:
#	src/main/kotlin/net/ccbluex/liquidbounce/LiquidBounce.kt

1zun4 committed:
feat: task managed deep learning initialization

24dedlaa commented:
So how do I turn the recorded json from the debug recorder module into a model?

sqlerrorthing commented:
> So how do I turn the recorded json from the debug recorder module into a model?

put your <file>.json into:
`<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`

24dedlaa commented:
> > So how do I turn the recorded json from the debug recorder module into a model?
> 
> put your .json into: `<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`

Ok, I made my model, played for a few minutes with debug recorder enabled with the minaraicombat mode. It wrote me where the model was saved as json. Now how do I make it so I can select it in killaura?

SuicidalEeper commented:
> > > So how do I turn the recorded json from the debug recorder module into a model?
> > 
> > 
> > put your .json into: `<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`
> 
> Ok, I made my model, played for a few minutes with debug recorder enabled with the minaraicombat mode. It wrote me where the model was saved as json. Now how do I make it so I can select it in killaura?

.models create name then  it's under rotation modes and you need both minaraicombat and minaraitrainer with like 10000 samples each

1zun4 commented:
75K Model with Sigmoid Correction Test
https://www.youtube.com/watch?v=6WbdAtx_l5w

CaptainBoggle commented:
@1zun4 Sorry for the ping, just had an idea and was wondering if it seems viable/makes sense at all to you:
Instead of recording combat data of the current player, could we instead record data from other players (even multiple simultaneously), such that we could simply AFK in spectator mode on a high-tier pvp server rather than relying on either the pretrained model (which could potentially be detectable at some point) or the users own pvp skill lol. This would also give us a lot more data to play with in a reasonable amount of time, as you could have alts recording across many servers/lobbies. The data won't be as precise due to the limitations of needing to interpret packets, and training on server-processed movements could result in some potentially strange behaviour such as the client never sending slightly out of bounds movements which the server would normally clamp etc. Anyway, just wanted to see if you had already considered this.

1zun4 commented:
> @1zun4 Sorry for the ping, just had an idea and was wondering if it seems viable/makes sense at all to you: Instead of recording combat data of the current player, could we instead record data from other players (even multiple simultaneously), such that we could simply AFK in spectator mode on a high-tier pvp server rather than relying on either the pretrained model (which could potentially be detectable at some point) or the users own pvp skill lol. This would also give us a lot more data to play with in a reasonable amount of time, as you could have alts recording across many servers/lobbies. The data won't be as precise due to the limitations of needing to interpret packets, and training on server-processed movements could result in some potentially strange behaviour such as the client never sending slightly out of bounds movements which the server would normally clamp etc. Anyway, just wanted to see if you had already considered this.

I have already tried it but it is a bit more problematic, because we basically need to detect if someone is in combat. If we start having samples where e.g. someone is only running away instead of being in combat, the model would be badly impacted. Also it would be better if we are the server instead of trying to record data from other players while being in client. I tried this on Hypixel Pit and did not manage to find a good way to do that.

With a lot of time you could of course do this, but at the moment it's much easier to record the combat data yourself.
